File: 08_greedy_algorithms/golang/01_set_covering.go
package main

import "fmt"

func main() {
	states_needed := []string{"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}

	var stations = make(map[string][]string)
	stations["kone"] = []string{"id", "nv", "ut"}
	stations["ktwo"] = []string{"wa", "id", "mt"}
	stations["kthree"] = []string{"or", "nv", "ca"}
	stations["kfour"] = []string{"nv", "ut"}
	stations["kfive"] = []string{"ca", "az"}

	station_key := []string{"kone", "ktwo", "kthree", "kfour", "kfive"}

	var final_stations []string

	for len(states_needed) > 0 {
		var best_station string
		var states_covered []string

		for _, station := range station_key {
			states := stations[station]
			var covered = equaldata(states_needed, states)
			if len(covered) > len(states_covered) {
				best_station = station
				states_covered = covered
			}
		}
		states_needed = removedata(states_needed, states_covered)
		final_stations = append(final_stations, best_station)

	}

	fmt.Println(final_stations)
}

func equaldata(states_needed []string, states []string) []string {
	var covered []string

	for _, state_needed := range states_needed {
		for _, state := range states {
			if state_needed == state {
				covered = append(covered, state_needed)
			}
		}
	}
	return covered
}

func removedata(states_needed []string, states_covered []string) []string {
	for _, state_covered := range states_covered {
		states_needed = remove(states_needed, state_covered)
	}
	return states_needed
}

func remove(states_needed []string, state_covered string) []string {
	for i, state_needed := range states_needed {
		if state_covered == state_needed {
			return append(states_needed[:i], states_needed[i+1:]...)
		}
	}
	return states_needed
}

File: 08_greedy_algorithms/elixir/01_set_covering.exs
defmodule SetCovering do
  def find_stations(states_needed, stations) do
    stations_list = for station <- stations, do: station
    do_find_stations(states_needed, MapSet.new(), stations_list)
  end

  defp do_find_stations(states_needed, final_stations, stations) do
    cond do
      MapSet.size(states_needed) == 0 ->
        final_stations

      true ->
        {station, states_covered} = find_best_station(states_needed, stations)

        do_find_stations(
          MapSet.difference(states_needed, states_covered),
          MapSet.put(final_stations, station),
          stations
        )
    end
  end

  defp find_best_station(states_needed, [{station_name, station_states} | []]) do
    {station_name, MapSet.intersection(states_needed, station_states)}
  end

  defp find_best_station(states_needed, [{station_name, station_states} | tail]) do
    (best_station = {_, best_states_covered}) = find_best_station(states_needed, tail)

    states_covered = MapSet.intersection(states_needed, station_states)

    if MapSet.size(states_covered) > MapSet.size(best_states_covered),
      do: {station_name, states_covered},
      else: best_station
  end
end

states_needed = MapSet.new(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"])

stations = %{
  "kone" => MapSet.new(["id", "nv", "ut"]),
  "ktwo" => MapSet.new(["wa", "id", "mt"]),
  "kthree" => MapSet.new(["or", "nv", "ca"]),
  "kfour" => MapSet.new(["nv", "ut"]),
  "kfive" => MapSet.new(["ca", "az"])
}

IO.inspect(SetCovering.find_stations(states_needed, stations))
